### PR TITLE
GLTF Animation Target Override Support

### PR DESCRIPTION
**Pull request by MackeyK24**

### CODE REVIEW

File: loaders/src/glTF/2.0/glTFLoader.ts [hunk 1]
```diff
@@ -1149,7 +1149,7 @@ export class GLTFLoader implements IGLTFLoader {
         });
     }
 
-    private _loadAnimationChannelAsync(context: string, animationContext: string, animation: IAnimation, channel: IAnimationChannel, babylonAnimationGroup: AnimationGroup): Promise<void> {
+    private _loadAnimationChannelAsync(context: string, animationContext: string, animation: IAnimation, channel: IAnimationChannel, babylonAnimationGroup: AnimationGroup, animationTargetOverride: Nullable<IAnimatable> = null): Promise<void> {
```
Comment (bghgary) [inline loaders/src/glTF/2.0/glTFLoader.ts]: @MackeyK24 I'm not sure about this change. I'm assuming you will be calling this private function from your code. Calling private functions is not ideal. Perhaps this function should be public underscore?
```suggestion
    public _loadAnimationChannelAsync(context: string, animationContext: string, animation: IAnimation, channel: IAnimationChannel, babylonAnimationGroup: AnimationGroup, animationTargetOverride: Nullable<IAnimatable> = null): Promise<void> {
```

File: loaders/src/glTF/2.0/glTFLoader.ts [hunk 2]
```diff
@@ -1289,8 +1289,13 @@ export class GLTFLoader implements IGLTFLoader {
                 const babylonAnimation = new Animation(animationName, targetPath, 1, animationType);
                 babylonAnimation.setKeys(keys);
 
-                targetNode._babylonTransformNode!.animations.push(babylonAnimation);
-                babylonAnimationGroup.addTargetedAnimation(babylonAnimation, targetNode._babylonTransformNode!);
+                if (animationTargetOverride != null && animationTargetOverride.animations != null) {
```
Comment (bghgary) [inline loaders/src/glTF/2.0/glTFLoader.ts]: The animationTargetOverride only affects node animations? Is there any consideration for morph targets?

Comment (MackeyK24) [reply loaders/src/glTF/2.0/glTFLoader.ts]: Nope... No Morph Targets.

This is going to be used for custom skinning. Where i create and use the actual Babylon Skeleton and bones... The default GLTF importers assign those animations and targets t the transform nodes of the skin... But i need to attach them to the actual bones. Here is how i am using it in my custom loadAnimationAsync in my extension:

```
    /** @hidden */
    public loadAnimationAsync(context: string, animation: BABYLON.GLTF2.IAnimation): Promise<BABYLON.AnimationGroup> {
        if (this._parseScene === true && animation.extras != null && animation.extras.metadata != null) {
            //console.warn("CVTOOLS: LoadAnimationAsync: " + animation.name);
            const loader:any = this._loader;
            const xanimation:any = animation;

            //////////////////////////////////////////////////////
            // Mackey Animation Modifications
            //////////////////////////////////////////////////////
            const metadata:any = animation.extras.metadata;
            const skeletonid:number = (metadata.skeletonid != null) ? metadata.skeletonid : null;
            let skeletonmap:any = null;
            if (skeletonid != null) {
                const skeleton:BABYLON.Skeleton = this._loader.babylonScene.getSkeletonById(<any>skeletonid);
                if (skeleton != null) {
                    if (skeleton.bones != null && skeleton.bones.length > 0) {
                        skeleton.bones.forEach((bone:BABYLON.Bone) => {
                            if (skeletonmap == null) skeletonmap = {};
                            skeletonmap[bone.id] = bone;
                        });
                    } else {
                        BABYLON.Tools.Warn("Skeleton '" + skeletonid.toString() + "' has no bones for animation: " + animation.name);
                    }
                } else {
                    BABYLON.Tools.Warn("Failed to locate skeleton '" + skeletonid.toString() + "' for animation: " + animation.name);
                }
            }
            //////////////////////////////////////////////////////

            const babylonAnimationGroup = new BABYLON.AnimationGroup(animation.name || `animation${xanimation.index}`, loader._babylonScene);
            xanimation._babylonAnimationGroup = babylonAnimationGroup;
    
            const promises = new Array<Promise<any>>();
    
            BABYLON.GLTF2.ArrayItem.Assign(xanimation.channels);
            BABYLON.GLTF2.ArrayItem.Assign(xanimation.samplers);
    
            for (const channel of animation.channels) 
            {
                if (channel.target.node != null && channel.target.node >= 0) {
                    //////////////////////////////////////////////////////
                    // Mackey Animation Modifications
                    //////////////////////////////////////////////////////
                    let animationTargetOverride:BABYLON.Bone = null;
                    if (skeletonmap != null && channel.extras != null && channel.extras.metadata != null) {
                        const cmetadata:any = channel.extras.metadata;
                        if (cmetadata.type != null && cmetadata.type === "bone" && cmetadata.targetid != null && cmetadata.targetid !== "") {
                            animationTargetOverride = skeletonmap[cmetadata.targetid];
                            if (animationTargetOverride != null) {
                                if (animationTargetOverride.animations == null) {
                                    animationTargetOverride.animations = [];
                                }
                            } else {
                                BABYLON.Tools.Warn("Failed to map bone '" + cmetadata.targetid + "' on skeleton '" + skeletonid.toString() + "' for animation: " + animation.name);
                            }
                        }
                    }
                    const xchannel:any = channel;
                    promises.push(loader._loadAnimationChannelAsync(`${context}/channels/${xchannel.index}`, context, animation, channel, babylonAnimationGroup, animationTargetOverride));
                    //////////////////////////////////////////////////////
                }
            }

            //////////////////////////////////////////////////////
            // Mackey Animation Modifications
            //////////////////////////////////////////////////////
            return Promise.all(promises).then(() => {
                const metadata:any = {};
                metadata.unity = animation.extras.metadata;
                (<any>babylonAnimationGroup).metadata = metadata;
                babylonAnimationGroup.normalize(0);
                return babylonAnimationGroup;
            });
            //////////////////////////////////////////////////////
        } else {
            return null; // Not Handled
        }
    }
```

Then my models that have a actual babylon skeleton will end up the animation on its bones (Like it was intended in the exporter... GLTF assigns the animation to transform nodes instead)

And the animation group with control the skeleton when playing the animation using Animation Groups

Works like a charm :)

File: loaders/src/glTF/2.0/glTFLoader.ts [hunk 3]
```diff
@@ -1149,7 +1149,17 @@ export class GLTFLoader implements IGLTFLoader {
         });
     }
 
-    private _loadAnimationChannelAsync(context: string, animationContext: string, animation: IAnimation, channel: IAnimationChannel, babylonAnimationGroup: AnimationGroup): Promise<void> {
+    /**
+     * Loads a glTF animation channel.
```
Comment (bghgary) [inline loaders/src/glTF/2.0/glTFLoader.ts]:
```suggestion
     * @hidden Loads a glTF animation channel.
```

**Comment by deltakosh** [comment]
Waiting for the CI to finish and I'll merge

**Pull request merged**